### PR TITLE
Bake Hacker News discussions into entries at build time

### PR DESCRIPTION
**Pull request by llun**
## What

Hacker News feeds (the official RSS, hnrss, mirrors) publish entries whose entire content is a **Comments** link, so the reader has nothing to show. This bakes the actual discussion into the entry while the action builds the site:

- **Detection** — `parseRss` now carries the RSS `<comments>` element on the entry (both the official HN RSS and hnrss put the item page there; the entry link is the article). Entries that link directly to the item page are also caught, as a fallback.
- **Fetching** — one request per entry to the HN Algolia API (`/api/v1/items/:id`), which returns the whole comment tree; the Firebase API would take one request per comment. Sequential, 10s timeout, and a failed fetch just keeps the original content — the build never breaks over it.
- **Rendering** — the feed's **Comments link stays at the top**; below it the story text (Ask HN self-posts) and the **top 20 comments, nested 3 levels deep**, with author profile links and date permalinks, ending in a "More comments on Hacker News" link when the caps cut the thread. Dead/deleted comments are skipped.
- **Sanitization** — the generated markup goes through the same `mapContentUrls` pass as feed content (relative comment links resolve against the item page, not the article). `class` is now allowed on `div`/`p`, restricted via `allowedClasses` to the `hn-*` classes, so the thread markup survives both the enrichment pass and the media store's rewrite; feed content classes keep being discarded.
- **Reader** — no component changes; a few `feeds-prose` rules style the thread (indent border for replies, muted meta line).

Verified end to end against the live HN RSS + Algolia API, and covered by 13 new tests (detection, caps, nesting, sanitization, escaping, failure fallback, class survival through the media rewrite pass). Full suite: 165 passing.

## Notes

- The enrichment runs inside `createLocalizingFeedLoader`, so both `files` and `sqlite` storage get it.
- Comments are a build-time snapshot, refreshed on every scheduled run — same freshness model as the feeds themselves.

### CODE REVIEW

File: action/feeds/hackernews.ts [hunk 1]
```diff
@@ -0,0 +1,204 @@
...
+    : ''
+  const meta = `<p class="hn-comment-meta">${authorLink}${dateLink}</p>`
+
+  const children = comment.children ?? []
+  let childrenHtml = ''
+  let truncated = false
+  if (depth >= MAX_COMMENT_DEPTH) {
+    truncated = children.length > 0
+  } else {
+    const rendered = children.map((child) => renderComment(child, depth + 1))
```
Comment (llun) [inline action/feeds/hackernews.ts]: **major — the caps don't bound output size.** Depth is capped at 3 and top-level count at 20, but replies per comment are unbounded, so total output is 20×N×M — a synthetic 20×500×50 thread produces a ~153 MB entry that gets stored and shipped to every reader. HN threads with hundreds of replies per comment are routine, and the input is fully attacker-influenceable. Related: `fetchItem` buffers the whole Algolia response with no size cap. Suggest a total rendered-comment budget (e.g. ~100) that sets `truncated` when exhausted, plus a byte cap on the response body.

**Review by llun — COMMENTED** [review]
Sub-agent review loop, round 1: three review agents covered correctness, security, and integration/tests of this diff. Consolidated findings are attached inline; fixes (or explicit won't-fix justifications) follow in this PR.

File: action/feeds/hackernews.ts [hunk 2]
```diff
@@ -0,0 +1,204 @@
...
+      if (!addition) {
+        entries.push(entry)
+        continue
+      }
+      // Comment HTML lives on the item page, so that is its document base --
+      // not the entry link, which for a HN feed is the article itself.
+      const itemUrl = `https://news.ycombinator.com/item?id=${id}`
+      const content =
+        entry.content +
+        mapContentUrls(addition, (url, target) =>
```
Comment (llun) [inline action/feeds/hackernews.ts]: **major — comment `` becomes media-store download input (SSRF + image laundering).** Comment `text` is third-party HTML and `img` is an allowed tag, so any HN commenter can plant `` or `https://attacker.example/x.png`; enrichment runs before `localizeSite`, so `collectDownloadableMediaUrls` picks those up and the build runner fetches them — blind SSRF on self-hosted runners, and any image-typed response gets republished under the feed site's own domain. Before this PR only operator-chosen feeds controlled fetched URLs. HN itself renders no images in comments, so dropping `img` from this sanitize pass loses nothing and closes the vector.

File: action/feeds/hackernews.ts [hunk 3]
```diff
@@ -0,0 +1,204 @@
...
+function renderComment(comment: AlgoliaItem, depth: number): Rendered {
+  // Dead and deleted comments have no author or text; skip them silently.
+  if (!comment.author || !comment.text) return { html: '', truncated: false }
+
+  const author = escapeHtml(comment.author)
+  const authorLink = `<a href="https://news.ycombinator.com/user?id=${encodeURIComponent(
+    comment.author
+  )}">${author}</a>`
+  const dateLink = comment.created_at_i
+    ? ` · <a href="https://news.ycombinator.com/item?id=${comment.id}">${format(
```
Comment (llun) [inline action/feeds/hackernews.ts]: **major — an invalid `created_at_i` drops the whole thread.** `format(new Date(created_at_i * 1000), 'PP')` throws `RangeError: Invalid time value` on out-of-range input (JSON happily parses `1e999` to `Infinity`); the throw lands in the per-entry catch, so one poisoned comment discards the entry's entire thread. Validate (`Number.isFinite` + valid `Date`) before formatting and omit the date link on bad input.

File: action/feeds/hackernews.ts [hunk 4]
```diff
@@ -0,0 +1,204 @@
...
+}
+
+/**
+ * Renders one comment and its replies. Replies nest inside a bordered block up
+ * to MAX_COMMENT_DEPTH levels; a comment deeper than that still exists on the
+ * item page, which is what the truncated flag bubbles up for.
+ */
+function renderComment(comment: AlgoliaItem, depth: number): Rendered {
+  // Dead and deleted comments have no author or text; skip them silently.
+  if (!comment.author || !comment.text) return { html: '', truncated: false }
```
Comment (llun) [inline action/feeds/hackernews.ts]: **minor — dead comments silently vanish their live replies.** Algolia keeps live replies attached to dead/deleted comments; this early return drops the whole subtree and reports `truncated: false`, so no "More comments" link appears either. HN itself shows a `[deleted]` placeholder and keeps the replies. Render the children in the dead comment's place (no body/meta) instead of dropping them.

File: action/feeds/hackernews.ts [hunk 5]
```diff
@@ -0,0 +1,204 @@
...
+ * the item page.
+ *
+ * A failed fetch never fails the build: the entry keeps its original content.
+ */
+export async function enrichSiteWithHackerNewsComments(
+  site: Site,
+  fetchApi: typeof globalThis.fetch = globalThis.fetch
+): Promise<Site> {
+  const entries: Entry[] = []
+  for (const entry of site.entries) {
```
Comment (llun) [inline action/feeds/hackernews.ts:365]: **minor — no time budget across the sequential per-entry fetches.** Each entry gets a 10 s timeout; a feed of 30 HN entries can burn ~5 minutes if Algolia hangs, and this runs before `localizeSite`, whose `LOCALIZE_DEADLINE_MS` is fixed at store creation — so slow enrichment eats the media download budget (media then degrades to remote URLs). Add a shared enrichment deadline (e.g. 60 s per site) and skip remaining fetches once exceeded, mirroring the media store.

File: action/feeds/hackernews.ts [hunk 6]
```diff
@@ -0,0 +1,204 @@
...
+ * article), so that is checked first; the entry link covers feeds that point
+ * their entries straight at the item page.
+ */
+export function hackerNewsItemId(url?: string) {
+  if (!url) return null
+  try {
+    const parsed = new URL(url)
+    if (parsed.hostname !== 'news.ycombinator.com') return null
+    if (parsed.pathname !== '/item') return null
+    return parsed.searchParams.get('id')
```
Comment (llun) [inline action/feeds/hackernews.ts]: **minor — the id is returned unvalidated.** `?id=` returns `''`, which via `??` blocks the entry-link fallback and then fails the `!id` guard — an entry whose link *is* a valid item page gets skipped. Non-numeric ids also flow unencoded into the fetch path and `itemUrl`. Return null unless the id matches `/^\d+$/`.

File: action/feeds/hackernews.ts [hunk 7]
```diff
@@ -0,0 +1,204 @@
...
+  if (commentsHtml) {
+    parts.push(`<div class="hn-comments">${commentsHtml}</div>`)
+  }
+
+  const truncated =
+    topLevel.length > MAX_TOP_LEVEL_COMMENTS ||
+    rendered.some((comment) => comment.truncated)
+  if (truncated) {
+    parts.push(
+      `<p class="hn-more"><a href="https://news.ycombinator.com/item?id=${item.id}">More comments on Hacker News</a></p>`
```
Comment (llun) [inline action/feeds/hackernews.ts]: **nit — the More link trusts `item.id`.** If the response omits `id` (shape drift), this renders `item?id=undefined`. The request id is already in scope at the call site — pass it into `renderItem` and use it instead.

File: action/feeds/hackernews.test.ts [hunk 8]
```diff
@@ -0,0 +1,323 @@
...
+test('#enrichSiteWithHackerNewsComments leaves non-HN entries alone', async (t) => {
+  const fetchStub = sinon.stub()
+  const site = createSite('https://example.com/item?id=1', undefined)
+  const enriched = await enrichSiteWithHackerNewsComments(
+    site,
+    fetchStub as any
+  )
+  t.is(fetchStub.callCount, 0)
+  t.deepEqual(enriched, site)
+})
```
Comment (llun) [inline action/feeds/hackernews.test.ts:85]: **nit — fallback semantics aren't pinned.** The fallback test covers `comments` absent + link is the item page, but not `comments` set to a *non-HN* URL with the link being the item page — which is exactly the case that distinguishes `??` from a buggy truthiness check. One more fixture would pin it.

File: action/feeds/parsers#parseRss.test.ts [hunk 9]
```diff
@@ -28,6 +28,8 @@ test('#parseAtom returns site information with empty string for fields that does
     link: firstEntry.link.join('').trim(),
     date: new Date('2021-02-08T10:05:48Z').getTime(),
     content: firstEntry['content:encoded'].join('').trim(),
-    author: 'icez'
+    author: 'icez',
+    comments:
```
Comment (llun) [inline action/feeds/parsers#parseRss.test.ts:32]: **nit — `comments` extraction edge cases are untested.** The new assertion covers only the happy path. Worth adding: missing `<comments>` element → `comments` undefined/absent, and a relative `<comments>` URL resolved against the site link.

File: action/feeds/parsers#parseRss.test.ts [hunk 10]
```diff
@@ -28,6 +28,8 @@ test('#parseAtom returns site information with empty string for fields that does
     link: firstEntry.link.join('').trim(),
     date: new Date('2021-02-08T10:05:48Z').getTime(),
     content: firstEntry['content:encoded'].join('').trim(),
```
Comment (llun) [inline action/feeds/parsers#parseRss.test.ts:30]: **nit — pre-existing misnomer this PR touches:** the test is named `#parseAtom returns site information…` but exercises `parseRss` on `rss1.xml`. Since this PR edits the assertion block, rename it to `#parseRss …` here.

File: action/feeds/parsers.ts [hunk 11]
```diff
@@ -14,6 +14,10 @@ export interface Entry {
   date: number
   content: string
   author: string
+  // The RSS <comments> URL, when the feed offers one. Hacker News feeds link
```
Comment (llun) [inline action/feeds/parsers.ts:17]: **nit — storage asymmetry worth documenting.** `comments` is spread into `EntryData` (files storage persists it in every published entry JSON) while sqlite's `insertEntry` has no column for it and drops it. No functional consequence today — nothing reads it back — but a future feature that reads `comments` from storage would behave differently per `storageType`. A note in this doc comment that it's build-time-only data (and files-only persistence) would defuse the trap.

File: action/feeds/index.ts [hunk 12]
```diff
@@ -41,7 +42,10 @@ async function createLocalizingFeedLoader(githubActionPath: string) {
   const feedLoader = async (title: string, url: string) => {
     const site: Site | null = await loadFeed(title, url)
     if (!site) return null
-    return store.localizeSite(site)
+    // HN entries carry only a "Comments" link, so the discussion is fetched
```
Comment (llun) [inline action/feeds/index.ts]: **nit — the wiring itself is untested.** No test asserts enrichment runs before `localizeSite`; deleting the hook or swapping the order fails no test. Noting for the record — the pipeline is thin and both halves are tested individually, so a won't-fix with justification is reasonable here.

File: action/feeds/hackernews.ts (hunk 1 above)
Comment (llun) [reply action/feeds/hackernews.ts]: Fixed in bb8d201. A shared budget () is now threaded through / on top of the 20/3 caps; exhausting it sets  so the More-comments link appears.  also reads the body as a stream with a 5 MB cap (), aborting the socket on oversize. Covered by "bounds the total comments rendered" (exactly 100  divs from a 400-comment input) and "keeps the original content when the response is too large".

File: action/feeds/hackernews.ts (hunk 2 above)
Comment (llun) [reply action/feeds/hackernews.ts]: Fixed in bb8d201. The addition is now sanitized with  — the stored-content tag set minus  — via a new optional  parameter on . HN renders no images in comments, so nothing is lost, and comment HTML can no longer hand URLs to the media downloader. Covered by "strips images from comment html", which also asserts  over the enriched content is empty.

File: action/feeds/hackernews.ts (hunk 3 above)
Comment (llun) [reply action/feeds/hackernews.ts]: Fixed in bb8d201.  guards with  and a NaN  check before calling date-fns; a bad timestamp now just omits the date link while the rest of the thread renders. Covered by "tolerates an invalid comment date", which feeds a raw  JSON literal (parses to ) and asserts the healthy sibling comment survives.

File: action/feeds/hackernews.ts (hunk 4 above)
Comment (llun) [reply action/feeds/hackernews.ts]: Fixed in bb8d201. A dead/deleted comment now renders its children in its place (wrapped in , still subject to the depth cap and budget), matching HN's  behavior, instead of vanishing the subtree. Covered by "renders live replies of a dead comment".

File: action/feeds/hackernews.ts (hunk 5 above)
Comment (llun) [reply action/feeds/hackernews.ts:365]: Fixed in bb8d201.  now works against a per-site deadline (); once exceeded, remaining entries keep their original content and enrichment moves on, so a stalled Algolia can no longer eat the media store's . Covered by "stops enriching after the deadline" with a stubbed clock.

File: action/feeds/hackernews.ts (hunk 6 above)
Comment (llun) [reply action/feeds/hackernews.ts]: Fixed in bb8d201.  now returns null unless the id matches  — an empty id no longer blocks the entry-link fallback, and a non-numeric one never reaches the fetch path or . Detection test extended with , , and a traversal-shaped id.

File: action/feeds/hackernews.ts (hunk 7 above)
Comment (llun) [reply action/feeds/hackernews.ts]: Fixed in bb8d201.  now takes the request id and builds the More-comments link from it, so response shape drift can't produce . Covered by "keys the more link on the request id" (response without an uid=501(llun) gid=20(staff) groups=20(staff),12(everyone),61(localaccounts),79(_appserverusr),80(admin),81(_appserveradm),98(_lpadmin),701(com.apple.sharepoint.group.1),33(_appstore),100(_lpoperator),204(_developer),250(_analyticsusers),395(com.apple.access_ftp),398(com.apple.access_screensharing),399(com.apple.access_ssh),400(com.apple.access_remote_ae) field).

File: action/feeds/hackernews.test.ts (hunk 8 above)
Comment (llun) [reply action/feeds/hackernews.test.ts:85]: Added in bb8d201: "falls back when the comments link is not a HN item" —  set to a non-HN URL with the entry link being the item page, pinning the  semantics.

File: action/feeds/parsers#parseRss.test.ts (hunk 9 above)
Comment (llun) [reply action/feeds/parsers#parseRss.test.ts:32]: Added in bb8d201: missing  element →  undefined and absent from the stored JSON ( round-trip), and a relative  resolved against the site link.

File: action/feeds/parsers#parseRss.test.ts (hunk 10 above)
Comment (llun) [reply action/feeds/parsers#parseRss.test.ts:30]: Renamed in bb8d201: the test is now .

File: action/feeds/parsers.ts (hunk 11 above)
Comment (llun) [reply action/feeds/parsers.ts:17]: Addressed with a doc note in bb8d201 rather than code:  now documents that it is build-time-only data, persisted by files storage and dropped by sqlite, and that nothing reads it back. Stripping it from  would add code for no functional benefit today.

File: action/feeds/index.ts (hunk 12 above)
Comment (llun) [reply action/feeds/index.ts]: Won't fix, with justification: the wiring is three lines composing two individually tested halves ( is covered by 20 tests, the media store by its own suite), and the composition was verified end to end against the live HN RSS + Algolia API. Testing it would require exporting  and stubbing the filesystem restore and media store — disproportionate to the risk it covers.

File: action/feeds/hackernews.ts [hunk 13]
```diff
@@ -0,0 +1,323 @@
...
+
+function renderChildren(
+  children: AlgoliaItem[],
+  depth: number,
+  budget: Budget
+): Rendered {
+  const parts: string[] = []
+  let truncated = false
+  for (const child of children) {
+    if (budget.remaining <= 0) {
```
Comment (llun) [inline action/feeds/hackernews.ts:175]: **minor (round 2) — `truncated` false positives: the More link can render when nothing *visible* was cut.** Four confirmed paths, all involving dead comments: budget exhausted with only dead leaves remaining; live comment at the depth cap whose children are all dead; dead comment at the depth cap with all-dead children (unconditional `truncated: true`); and 21+ dead-leaf top-level comments (`topLevel.length > MAX` counts dead ones). Gate the flag on liveness — an iterative `hasLiveComment` (the tree is unbounded-depth network input, so no recursion) — in all four paths.

File: action/feeds/hackernews.ts [hunk 14]
```diff
@@ -0,0 +1,323 @@
...
+    const rendered = renderChildren(children, depth + 1, budget)
+    childrenHtml = rendered.html
+    truncated = rendered.truncated
+  }
+  const childrenBlock = childrenHtml
+    ? `<div class="hn-comment-children">${childrenHtml}</div>`
+    : ''
+
+  return {
+    html: `<div class="hn-comment">${meta}<div class="hn-comment-body">${comment.text}</div>${childrenBlock}</div>`,
```
Comment (llun) [inline action/feeds/hackernews.ts]: **minor (round 2) — commenter HTML is sanitized in the same pass as the chrome, so it can wear the `hn-*` classes.** `<p class="hn-more"><a href="https://evil.example">More comments…</a></p>` as comment text survives intact (verified), masquerading as reader chrome. Latent today (HN/Algolia emit no class attributes in `text`), but this module treats `text` as untrusted everywhere else. Sanitize `comment.text`/`item.text` separately first, with classes fully disabled, then embed.

File: action/feeds/hackernews.ts [hunk 15]
```diff
@@ -0,0 +1,323 @@
...
+  type Site
+} from './parsers'
+
+const ALGOLIA_ITEM_API = 'https://hn.algolia.com/api/v1/items'
+const FETCH_TIMEOUT_MS = 10_000
+const MAX_RESPONSE_BYTES = 5 * 1024 * 1024
+// One budget per site, so a slow Algolia cannot eat into the media store's
+// localization deadline -- that clock starts when the loader is created,
+// before any enrichment runs.
+const ENRICHMENT_DEADLINE_MS = 60_000
```
Comment (llun) [inline action/feeds/hackernews.ts:20]: **minor (round 2) — the deadline claim is overstated in aggregate.** The budget is per *site*, but feeds load serially and the media store's 10-minute clock starts at loader creation — N HN feeds can still eat N×60s of it. Make the deadline per *run* (shared across sites), e.g. a `createHackerNewsEnricher` factory mirroring `createMediaStore`.

File: action/feeds/hackernews.ts [hunk 16]
```diff
@@ -0,0 +1,323 @@
...
+
+  budget.remaining--
+
+  const author = escapeHtml(comment.author)
+  const authorLink = `<a href="https://news.ycombinator.com/user?id=${encodeURIComponent(
+    comment.author
+  )}">${author}</a>`
+  // The id comes from the network; only a positive integer is interpolated
+  // into the permalink.
+  const id = Number.isInteger(comment.id) && comment.id > 0 ? comment.id : null
```
Comment (llun) [inline action/feeds/hackernews.ts]: **nit (round 2) — `Number.isInteger(1e21)` is true**, so a drifted huge `comment.id` serializes into the permalink as `item?id=1e+21` (verified). No injection (charset is `[0-9e+]`, plus the sanitize re-pass), but the link is nonsense. Use `Number.isSafeInteger`.

File: action/feeds/hackernews.ts [hunk 17]
```diff
@@ -0,0 +1,323 @@
...
+    .replace(/</g, '&lt;')
+    .replace(/>/g, '&gt;')
+    .replace(/"/g, '&quot;')
+}
+
+function formatCommentDate(timestamp?: number) {
+  // Untrusted input: JSON parses a literal like 1e999 to Infinity, and
+  // date-fns format throws a RangeError on an invalid Date, which the
+  // per-entry catch would turn into the whole thread being dropped.
+  if (!timestamp || !Number.isFinite(timestamp)) return ''
```
Comment (llun) [inline action/feeds/hackernews.ts]: **nit (round 2) — negative `created_at_i` passes the finite check and renders a 1969 date.** Reject `timestamp < 0` too.

File: action/feeds/parsers.ts [hunk 18]
```diff
@@ -167,13 +191,19 @@ export const ENTRY_CONTENT_SANITIZE_OPTIONS: sanitizeHtml.IOptions = {
...
  */
 export function mapContentUrls(
   content: string,
-  mapUrl: (url: string, target: UrlTarget) => string
+  mapUrl: (url: string, target: UrlTarget) => string,
+  allowedTags?: string[]
 ) {
   return sanitizeHtml(content, {
     ...ENTRY_CONTENT_SANITIZE_OPTIONS,
+    ...(allowedTags ? { allowedTags } : {}),
```
Comment (llun) [inline action/feeds/parsers.ts]: **nit (round 2) — `allowedTags ? … : …` treats `[]` as truthy**: a future caller passing an empty array would strip every tag instead of getting defaults. Use `allowedTags !== undefined`.

File: action/feeds/hackernews.test.ts [hunk 19]
```diff
@@ -0,0 +1,511 @@
...
+    await enrichSiteWithHackerNewsComments(
+      createSite(HN_ITEM_LINK),
+      fetchStub as any
+    )
+  ).entries
+  t.true(entry.content.includes('<p>reply to a dead comment</p>'))
+  t.false(entry.content.includes('user40003'))
+})
+
+test('#enrichSiteWithHackerNewsComments bounds the total comments rendered', async (t) => {
```
Comment (llun) [inline action/feeds/hackernews.test.ts:424]: **nit (round 2) — test gaps around the new fixes:** no case for "budget exhausted with nothing visible cut → no More link", none for the declared-`content-length` early-throw path (only the streaming path is exercised), and no boundary case at `now() === deadline`.

File: action/feeds/hackernews.ts [hunk 20]
```diff
@@ -0,0 +1,323 @@
...
+  comment: AlgoliaItem,
+  depth: number,
+  budget: Budget
+): Rendered {
+  const children = comment.children ?? []
+
+  // Dead and deleted comments have no author or text. HN keeps their live
+  // replies under a [deleted] marker, so render the children in the dead
+  // comment's place rather than vanishing the whole subtree.
+  if (!isLiveComment(comment)) {
```
Comment (llun) [inline action/feeds/hackernews.ts:206]: **nit (round 2) — two presentation/logging nits:** (a) a dead comment's children render inside a bordered indent with no visible parent — HN shows a `[deleted]` marker for context, worth mirroring; (b) the post-deadline skip logs one `console.error` per remaining entry, which is log spam on large feeds — log once per site.

File: action/feeds/hackernews.ts [hunk 21]
```diff
@@ -0,0 +1,323 @@
...
+// The depth and top-level caps alone do not bound reply breadth (20 comments
+// can each carry hundreds of replies), so the total is capped too.
+const MAX_TOTAL_COMMENTS = 100
+
+// HN renders no images in comments, and enrichment runs before the media
+// store -- an img left in here would hand URLs chosen by any commenter to the
+// downloader and republish them from this site's own origin.
+const COMMENT_ALLOWED_TAGS = (
+  (ENTRY_CONTENT_SANITIZE_OPTIONS.allowedTags as string[]) || []
+).filter((tag) => tag !== 'img')
```
Comment (llun) [inline action/feeds/hackernews.ts]: **minor (round 2) — `COMMENT_ALLOWED_TAGS` is derived by exclusion, not by property.** Correct today (verified: no surviving tag can carry a media URL), but `filter((tag) => tag !== 'img')` silently re-inherits any future media-carrying tag added to the shared sanitize options (e.g. `video` with `src`), re-opening the round-1 commenter→downloader path. Derive by property — drop every tag whose allowed attributes intersect `src`/`srcset` — and pin it with an invariant test.

File: action/feeds/parsers.ts [hunk 22]
```diff
@@ -137,7 +143,25 @@ export const ENTRY_CONTENT_SANITIZE_OPTIONS: sanitizeHtml.IOptions = {
...
+    // The Hacker News discussion markup (action/feeds/hackernews.ts) styles
+    // the thread through classes. Every pass over stored content goes through
+    // these options, so class has to survive here or the media store's rewrite
+    // would strip it back off.
+    div: ['class'],
+    p: ['class']
+  },
+  // Only the classes the HN markup uses; a feed's own classes keep being
+  // discarded as before.
+  allowedClasses: {
```
Comment (llun) [inline action/feeds/parsers.ts:156]: **nit (round 2) — `hn-story`, `hn-comment-body`, and `hn-more` have no matching rules in globals.css.** Harmless semantic hooks (and the spoofing surface is cosmetic-only: the styled rules set no position/z-index/pointer-events). Keep or trim; noting for the record.

**Review by llun — COMMENTED** [review]
Sub-agent review loop, round 2: re-review of the round-1 fixes (correctness + security agents). Findings inline; fixes follow.

File: action/feeds/hackernews.ts (hunk 13 above)
Comment (llun) [reply action/feeds/hackernews.ts:175]: Fixed in 8368785.  now means "something a reader could see was cut": an iterative  (explicit stack — the tree is unbounded-depth network input) gates the flag in all four paths — the budget break, both depth-cap branches, and the top-level slice remainder. Covered by "omits the more link when only dead comments were cut" (both the top-level and depth-cap variants).

File: action/feeds/hackernews.ts (hunk 14 above)
Comment (llun) [reply action/feeds/hackernews.ts]: Fixed in 8368785. Comment and story text are now pre-sanitized through  (shared options but / allowedAttributes plus ) before being embedded in the chrome, so commenter HTML cannot carry the  classes while the generated chrome keeps them. Covered by "strips commenter classes but keeps the chrome".

File: action/feeds/hackernews.ts (hunk 15 above)
Comment (llun) [reply action/feeds/hackernews.ts:20]: Fixed in 8368785. The deadline is now per *run*:  (mirroring ) computes it once and shares it across every site the loader processes, so N HN feeds can no longer burn N×60s of the media clock. Covered by "createHackerNewsEnricher shares one deadline across sites".

File: action/feeds/hackernews.ts (hunk 16 above)
Comment (llun) [reply action/feeds/hackernews.ts]: Fixed in 8368785 —  instead; an unsafe id now drops the permalink and the date renders as plain text. Covered by "renders no permalink for an unsafe comment id" (raw  JSON literal).

File: action/feeds/hackernews.ts (hunk 17 above)
Comment (llun) [reply action/feeds/hackernews.ts]: Fixed in 8368785 —  is rejected alongside the finite/NaN checks. Covered by "renders no date for a negative timestamp". (String timestamps keep being dropped — that matches the documented Algolia shape.)

File: action/feeds/parsers.ts (hunk 18 above)
Comment (llun) [reply action/feeds/parsers.ts]: Fixed in 8368785 — , with a comment explaining why truthiness would be wrong.

File: action/feeds/hackernews.test.ts (hunk 19 above)
Comment (llun) [reply action/feeds/hackernews.test.ts:424]: Added in 8368785: "omits the more link when only dead comments were cut", "refuses a declared oversize content-length before reading", and "enriches at the deadline boundary" ( still enriches).

File: action/feeds/hackernews.ts (hunk 20 above)
Comment (llun) [reply action/feeds/hackernews.ts:206]: Both fixed in 8368785: (a) dead comments with live replies now render a muted  meta line above the children, mirroring HN; (b) the deadline skip logs once per site ("Skip remaining Hacker News comments for <site>") instead of per entry.

File: action/feeds/hackernews.ts (hunk 21 above)
Comment (llun) [reply action/feeds/hackernews.ts]: Fixed in 8368785.  is now derived by property — any tag whose allowed attributes (tag-specific or ) intersect / is excluded — so a media-carrying tag added to the shared options later is dropped here automatically. Pinned by "COMMENT_ALLOWED_TAGS permits no media-carrying tag".

File: action/feeds/parsers.ts (hunk 22 above)
Comment (llun) [reply action/feeds/parsers.ts:156]: Won't fix, with justification: the unstyled classes are deliberate semantic hooks — they mark the story/body/more-link structure in the stored HTML and give the reader stable styling anchors at zero cost. The whitelist () is the safety boundary, and as noted the styled rules are cosmetic-only.

File: action/feeds/hackernews.ts [hunk 23]
```diff
@@ -0,0 +1,398 @@
...
+function renderChildren(
+  children: AlgoliaItem[],
+  depth: number,
+  budget: Budget
+): Rendered {
+  const parts: string[] = []
+  let truncated = false
+  for (const [index, child] of children.entries()) {
+    if (budget.remaining <= 0) {
+      truncated = hasLiveComment(children.slice(index))
```
Comment (llun) [inline action/feeds/hackernews.ts]: **major (round 3) — the budget-break path *overwrites* an accumulated `truncated = true`.** Round 1's `truncated = true` could only set; this assigns, so when the budget runs out mid-list and the remaining siblings are all dead, `hasLiveComment` returns false and wipes a `true` from an earlier sibling (e.g. a depth-capped live subtree). Verified: 100 comments rendered, live content cut, yet no More link — the exact false negative round 2 set out to eliminate. Accumulate instead: `truncated = truncated || hasLiveComment(children.slice(index))`, plus a regression test shaped as: earlier sibling truncated + budget break + all-dead remainder.

File: action/feeds/hackernews.ts [hunk 24]
```diff
@@ -0,0 +1,398 @@
...
+  id: number
+  author?: string | null
+  text?: string | null
+  created_at_i?: number
+  children?: AlgoliaItem[]
+}
+
+type LiveComment = AlgoliaItem & { author: string; text: string }
+
+function isLiveComment(comment: AlgoliaItem): comment is LiveComment {
```
Comment (llun) [inline action/feeds/hackernews.ts]: **minor (round 3) — malformed node shapes throw, and the per-entry catch turns one bad node into loss of the whole fetched thread.** Verified: `children: [null, <live>]` throws `TypeError` (`comment.children` on null), and an author with a lone surrogate (`"\ud800"`, which JSON.parse produces happily) makes `encodeURIComponent` throw `URIError`. Ids and dates are hardened against hostile input; these spots aren't. Fix: null/type-safe `isLiveComment`, skip non-object entries in `hasLiveComment`/`renderChildren`, `Array.isArray` guard on `children`, and normalize lone surrogates (e.g. `toWellFormed()`) before escaping/encoding the author.

File: action/feeds/hackernews.ts [hunk 25]
```diff
@@ -0,0 +1,398 @@
...
+ * tree is network input of unbounded depth. This is what the truncated flag
+ * means; counting dead leaves would render a More-comments link over a thread
+ * that shows nothing new.
+ */
+function hasLiveComment(items: AlgoliaItem[]): boolean {
+  const stack = [...items]
+  while (stack.length > 0) {
+    const item = stack.pop()!
+    if (isLiveComment(item)) return true
+    if (item.children) stack.push(...item.children)
```
Comment (llun) [inline action/feeds/hackernews.ts]: **nit (round 3) — `stack.push(...item.children)` hits V8's spread-argument limit on very wide nodes** (≳500k minimal dead children fit under the 5MB body cap; verified end-to-end: RangeError, thread dropped, build survives via the per-entry catch). Replace with a plain `for` loop push, which has no argument-count limit.

File: action/feeds/hackernews.ts [hunk 26]
```diff
@@ -0,0 +1,398 @@
...
+    topLevel.slice(0, MAX_TOP_LEVEL_COMMENTS),
+    1,
+    budget
+  )
+  if (rendered.html) {
+    parts.push(`<div class="hn-comments">${rendered.html}</div>`)
+  }
+
+  const truncated =
+    hasLiveComment(topLevel.slice(MAX_TOP_LEVEL_COMMENTS)) || rendered.truncated
```
Comment (llun) [inline action/feeds/hackernews.ts]: **nit (round 3) — operand order:** `hasLiveComment(topLevel.slice(...))` walks and copies the beyond-cap remainder even when `rendered.truncated` is already true. Swap: `rendered.truncated || hasLiveComment(...)`.

File: action/feeds/hackernews.ts [hunk 27]
```diff
@@ -0,0 +1,398 @@
...
+  ...ENTRY_CONTENT_SANITIZE_OPTIONS,
+  allowedAttributes: {
+    ...(ENTRY_CONTENT_SANITIZE_OPTIONS.allowedAttributes as Record<
+      string,
+      string[]
+    >),
+    div: [],
+    p: []
+  },
+  allowedClasses: {}
```
Comment (llun) [inline action/feeds/hackernews.ts:61]: **nit (round 3) — defense in depth: comment anchors keep `name` and `target`** (verified in stored output: `<a name="clobber0" target="_top">`). No exploitable sink today — the reader force-overwrites `target`/`rel` on every anchor and does no named-global lookups — but this PR expands those attributes' authors from feed publishers to arbitrary commenters. Restrict `a` to `['href']` in the pre-sanitize's allowedAttributes override.

**Review by llun — COMMENTED** [review]
Sub-agent review loop, round 3: re-review of the round-2 fixes. One major correctness bug in the new truncation logic, plus contained robustness gaps. Fixes follow.

File: action/feeds/hackernews.ts (hunk 23 above)
Comment (llun) [reply action/feeds/hackernews.ts]: Fixed in 26afd87 — the break path now accumulates (). Regression test added with exactly the shape you described: T1 depth-capped (truncated=true), T2's 96 replies exhaust the budget, T3–T20 all dead — asserts 100 rendered comments and the More link present.

File: action/feeds/hackernews.ts (hunk 24 above)
Comment (llun) [reply action/feeds/hackernews.ts]: Fixed in 26afd87.  is now null/type-safe,  and  skip non-object nodes,  goes through an  guard everywhere, and author names pass through  before escaping/encoding (lone surrogates become U+FFFD). Covered by "tolerates malformed nodes" (null entry + non-array children) and "normalizes lone surrogates in author names".

File: action/feeds/hackernews.ts (hunk 25 above)
Comment (llun) [reply action/feeds/hackernews.ts]: Fixed in 26afd87 — plain  loop push, with a comment noting why the spread form is unsafe at that width.

File: action/feeds/hackernews.ts (hunk 26 above)
Comment (llun) [reply action/feeds/hackernews.ts]: Fixed in 26afd87 — operands swapped so the remainder walk is skipped when the rendered portion is already truncated.

File: action/feeds/hackernews.ts (hunk 27 above)
Comment (llun) [reply action/feeds/hackernews.ts:61]: Fixed in 26afd87 — the pre-sanitize now restricts  to  for commenter HTML. Covered by "strips name and target from comment anchors".
